lib/epochs.js

```diff
@@ -99,141 +101,196 @@ module.exports = function Epochs(ssb) {
     })
   }
 
-  return {
-    getEpochs,
-
-    /**
-     * Get the epoch matching the currently picked write key for the group.
-     */
-    getPickedEpoch(groupId, opts) {
-      let lastKey
-
-      return pull(
-        opts?.live
-          ? ssb.box2.getGroupInfoUpdates(groupId)
-          : pull(ssb.box2.getGroupInfoUpdates(groupId), pull.take(1)),
-        pull.map((groupInfo) => {
-          // prettier-ignore
-          if (groupInfo.excluded) throw new Error("Can't get picked epoch for group we're excluded from")
-
-          return groupInfo
-        }),
-        // only pass forward group info if the writeKey is different to last time
-        pull.filter((groupInfo) => {
-          if (lastKey?.equals(groupInfo.writeKey.key)) return false
+  function getMembers(epochRootId, cb) {
+    if (cb === undefined) return p(getMembers)(epochRootId)
+
+    epochRootId = toMsgURI(epochRootId)
+    const added = new Set()
+    const toExclude = new Set()
+
+    pull(
+      getTangleUpdates.stream(ssb, 'members', epochRootId),
+      pull.filter((msg) => isAddMember(msg) || isExcludeMember(msg)),
+      pull.through((msg) => {
+        const { type, recps, excludes } = msg.value.content
+        if (type === 'group/add-member')
+          recps.slice(1).forEach((feedId) => added.add(feedId))
+        else return excludes.forEach((feedId) => toExclude.add(feedId))
+      }),
+      pull.collect((err) => {
+        if (err) return cb(clarify(err, 'Failed to resolve epoch membership'))
 
-          lastKey = groupInfo.writeKey.key
-          if (!lastKey) console.log('empty groupinfo?', groupInfo)
-          return true
-        }),
-        // find details about the newly picked epoch
-        pull.asyncMap((groupInfo, cb) => {
-          getEpochs(groupId, (err, epochs) => {
-            // prettier-ignore
-            if (err) return cb(clarify(err, "Error getting info about epochs when looking for picked epoch"))
-
-            console.log('got epochs but no err', epochs, 'groupInfo', groupInfo)
-
-            cb(
-              null,
-              epochs.find((epoch) =>
-                epoch.secret.equals(groupInfo.writeKey.key)
-              )
-            )
-          })
+        cb(null, {
+          added: [...added],
+          toExclude: [...toExclude],
         })
-      )
-    },
+      })
+    )
+  }
+  getMembers.stream = function getMembersStream(epochRootId, opts = {}) {
```

Comment:
changed `getMembers.stream` so that it consolidates updates emitted:
- loads current state and emits that
- only emits when there is new info discovered (and then emits the whole current state again)

Very happy with how this turned out. Wrote some tests over it too